rllib/evaluation/collectors/agent_collector.py

```diff
@@ -0,0 +1,545 @@
+                }
+            )
+
+        # Append data to existing buffers.
+        flattened = tree.flatten(init_obs)
+        for i, sub_obs in enumerate(flattened):
+            self.buffers[SampleBatch.OBS][i].append(sub_obs)
+        self.buffers[SampleBatch.AGENT_INDEX][0].append(agent_index)
+        self.buffers[SampleBatch.ENV_ID][0].append(env_id)
+        self.buffers[SampleBatch.T][0].append(t)
+        self.buffers[SampleBatch.EPS_ID][0].append(self.episode_id)
+        self.buffers[SampleBatch.UNROLL_ID][0].append(self.unroll_id)
+
+    def add_action_reward_next_obs(self, values: Dict[str, TensorType]) -> None:
+        """Adds the given dictionary (row) of values to the Agent's trajectory.
+
+        Args:
+            values: Data dict (interpreted as a single row) to be added to buffer.
+            Must contain keys:
+                SampleBatch.ACTIONS, REWARDS, DONES, and NEXT_OBS.
+        """
+        if self.unroll_id is None:
+            self.unroll_id = AgentCollector._next_unroll_id
+            AgentCollector._next_unroll_id += 1
+
+        # Next obs -> obs.
+        assert SampleBatch.OBS not in values
+        values[SampleBatch.OBS] = values[SampleBatch.NEXT_OBS]
+        del values[SampleBatch.NEXT_OBS]
+
+        # Make sure EPS_ID/UNROLL_ID stay the same for this agent.
+        if SampleBatch.EPS_ID in values:
+            assert values[SampleBatch.EPS_ID] == self.episode_id
+            del values[SampleBatch.EPS_ID]
+        self.buffers[SampleBatch.EPS_ID][0].append(self.episode_id)
+        if SampleBatch.UNROLL_ID in values:
+            assert values[SampleBatch.UNROLL_ID] == self.unroll_id
+            del values[SampleBatch.UNROLL_ID]
+        self.buffers[SampleBatch.UNROLL_ID][0].append(self.unroll_id)
+
+        for k, v in values.items():
+            if k not in self.buffers:
+                self._build_buffers(single_row=values)
+            # Do not flatten infos, state_out_ and (if configured) actions.
+            # Infos/state-outs may be structs that change from timestep to
+            # timestep.
+            if (
+                k == SampleBatch.INFOS
+                or k.startswith("state_out_")
+                or (k == SampleBatch.ACTIONS and not self.disable_action_flattening)
+            ):
+                self.buffers[k][0].append(v)
+            # Flatten all other columns.
+            else:
+                flattened = tree.flatten(v)
+                for i, sub_list in enumerate(self.buffers[k]):
+                    sub_list.append(flattened[i])
+
+        # In inference mode, we don't need to keep all of trajectory in memory
+        # we only need to keep the steps required. We can pop from the beginning to
+        # create room for new data.
+        if not self.training:
+            for k in self.buffers:
+                for sub_list in self.buffers[k]:
+                    if sub_list:
+                        sub_list.pop(0)
```

Comment:
we can simply pop(0) because we already built all the shift_before elements in self.buffers, right?

Reply:
no sub_list can still be empty. I had this bug and then I added the check.